### ISSUE TITLE
conv2d segfaults on an input with non-contiguous strides on size-1 spatial dims (reached from plain torch.compile via reinterpret_tensor)

### ISSUE THREAD

**Issue by laolvfan** [pytorch/pytorch#195394]
## 🐛 Describe the bug

`torch.compile(backend="inductor")` segfaults (SIGSEGV, core dumped) on a small `Conv2d` model that eager runs fine.

The underlying fault is in ATen's convolution: `torch.nn.functional.conv2d` segfaults when the input tensor has **non-contiguous strides on size-1 spatial dimensions** and a bias is supplied. Inductor reaches that state on its own — it rewrites the input with `reinterpret_tensor` before calling `extern_kernels.convolution`, so ordinary user code with no `as_strided` anywhere hits the crash.

## Reproduction — via `torch.compile` (no manual strides)

```python
import torch
import torch.nn as nn

class M(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(64, 1, kernel_size=(125, 1), stride=(2, 2), padding=(63, 63))
        self.linear = nn.Linear(64, 64, bias=False)

    def forward(self, x1, x2):
        return torch.relu(self.linear(self.conv(x1)) + x2)

torch.manual_seed(0)
m = M().eval()
x1 = torch.randn(100, 64, 1, 1)
x2 = torch.randn(100, 64, 1, 1)

with torch.no_grad():
    print("eager ok:", tuple(m(x1, x2).shape))          # (100, 64, 2, 64)
    torch.compile(m, backend="inductor")(x1, x2)        # Segmentation fault (core dumped)
```

Eager prints `(100, 64, 2, 64)`; the compiled call dies with SIGSEGV (shell reports exit code 139).

## Reproduction — pure eager, no `torch.compile`

The same crash without the compiler, using the exact strides Inductor produces:

```python
import torch

N, C, K, P = 100, 64, 125, 63
base = torch.randn(N, C, 1, 1)
x = torch.as_strided(base, (N, C, 1, 1), (C, 1, C, C))   # strides on the size-1 spatial dims
assert torch.equal(x, base)                              # elementwise identical to base

w = torch.randn(1, C, K, 1)
b = torch.randn(1)
torch.nn.functional.conv2d(x, w, b, stride=(2, 2), padding=(P, P))   # Segmentation fault
```

`x` and `base` are the same tensor elementwise — for size-1 dimensions the stride is not observable through indexing — but only the restrided one crashes.

## Where it comes from in the compiled path

`PYTHONFAULTHANDLER=1` puts the fault inside the generated kernel, at the convolution call:

```
File ".../torch/_ops.py", line 875 in __call__
File "/tmp/.../cjq3ha...py", line 146 in call
```

and line 146 of the generated code is:

```python
buf1 = extern_kernels.convolution(
    reinterpret_tensor(primals_3, (100, 64, 1, 1), (64, 1, 64, 64), 0),
    buf0, primals_2, stride=(2, 2), padding=(63, 63), dilation=(1, 1),
    transposed=False, output_padding=(0, 0), groups=1)
```

`primals_3` is asserted earlier as `(100, 64, 1, 1)` with stride `(64, 1, 1, 1)`; Inductor hands the convolution the same tensor restrided to `(64, 1, 64, 64)`.

## Conditions

- **Bias is required.** The same call with `bias=None` returns normally.
- **Batch size matters, irregularly.** With `C=64, K=125, P=63`:

  | N | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 100 |
  |---|---|---|---|---|---|---|---|---|---|----|-----|
  | result | ok | ok | **segv** | ok | **segv** | **segv** | ok | ok | **segv** | **segv** | **segv** |

  Each configuration is deterministic (3/3 identical runs), but whether a given N faults does not follow a clean rule — the signature of an out-of-bounds read whose landing address depends on the allocation layout.
- When it does not fault, the output is correct: comparing the restrided input against the contiguous one gives `maxdiff = 0`, so this is a crash, not silent corruption.

## Versions

Reproduces identically on:

- PyTorch `2.11.0`
- PyTorch `2.13.0`

CPU only, Linux x86_64, Python 3.11.

## Additional context

Checked against nearby reports; I do not believe this is a duplicate:

- **#184027** (`[inductor] torch.compile crashes on convolutions when weight is a plain tensor — stride prediction mismatch`) raises a clean `AssertionError` on CUDA and is triggered by passing a plain-tensor weight. This one is a SIGSEGV on CPU with an ordinary `nn.Conv2d`, and reproduces in eager with no compiler involved.
- **#183372** (`torch.compile Inductor crashes on conv2d with bool-cast input and flattened bias`) is a `LoweringException` at compile time, not a memory fault.

Found by differential testing over StarCoder-generated programs; two independently generated programs produced the same generated-code shape (`reinterpret_tensor((N, C, 1, 1), (C, 1, C, C))` into `extern_kernels.convolution`) and the same SIGSEGV, and the repros above are reduced from them.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

**Comment by HussainNizamani** [comment]
**Reproduced on two builds, and a defensive `.clone()` inside the compiled function doesn't survive either**

Confirmed on my end. I ran this on two separate builds on aarch64 CPU, a recent nightly wheel (`2.15.0.dev20260824+cpu`) and a local source build (`2.14.0a0+git9c98593`), and got identical results on both. 

Every version I tried triggers it, the MWE as written, plain `torch.compile(fn)` without fullgraph, and a realistic optimizer style pattern where an in place op applied to the returned tensor after the step silently changed the parameter as well.

I tried the obvious defensive fix, returning `updated.clone()` from inside the compiled function, and Inductor removes that clone too.

The output still aliases the input. 

Here are analysis from my AI:

>
> **What breaks.** In eager mode `fn` returns a tensor that is independent of `x`, so mutating the return value afterwards leaves `x` untouched. Under `torch.compile` the returned tensor is an alias of `x`, and any later in-place op on the output writes into the input. There is no error or warning at any point. A concrete consequence, using an optimizer-style caller that applies `new_param.mul_(0.9)` after the step: eager keeps `param = [10., 2.]`, compiled ends with `param = [9.0, 1.8]`.
>
> **Root cause.** Consistent with the diagnosis in the issue body. Functionalization produces `scatter -> copy_(input, scatter) -> return scatter`, and `should_reinplace_scatter()` treats the direct copy-back as profitable and reinplaces the scatter onto the input buffer without checking that the functional scatter result is still live as a graph output. Running the same MWE with `backend="aot_eager"` and `backend="eager"` gives correct, non-aliasing results, which localizes the bug to Inductor's reinplacing pass rather than functionalization or AOTAutograd. The clone experiment below suggests the surrounding buffer-reuse/copy-elimination then also removes a defensive clone whose source was reinplaced, though we did not trace that second interaction to a specific line.
>
> **Workarounds tested** (nightly wheel `2.15.0.dev20260824+cpu`, aarch64 CPU; each probe checks `torch._C._is_alias_of(x, out)` and whether mutating `out` corrupts `x`):
>
> | Probe | Output aliases input | Input corrupted |
> |---|---|---|
> | eager baseline | no | no |
> | `torch.compile` (Inductor, default) | **yes** | **yes** |
> | `torch.compile(fn)` without fullgraph | **yes** | **yes** |
> | `backend="aot_eager"` | no | no |
> | `backend="eager"` | no | no |
> | `return updated.clone()` inside the compiled function (Inductor) | **yes** | **yes** |
> | `compiled_fn(...)` then `.clone()` outside the compiled region | no | no |
>
> We also searched `torch._inductor.config` for a toggle covering this pass and found no config attribute matching `*inplace*` on this build, so there is no kill-switch style mitigation.
>
> **Why a workaround is not sufficient.** The only mitigation that works is cloning the result outside the compiled region, which requires every caller to already know about the bug. Code inside a compiled region cannot defend itself, since an explicit clone of the scatter result is eliminated. The failure is silent, changes the eager/compile aliasing contract, and shows up in ordinary patterns (update a buffer, return the result, mutate it later). That combination is what the `module: correctness (silent)` label exists for, and it argues for a liveness/alias check on the direct copy-back path in `should_reinplace_scatter()` rather than documentation or a user-side recipe.

**Comment by harsh-vsingh** [comment]
Reproduced independently. Confirmed root cause via gdb + mkldnn toggle.

**1. Minimal repro (no `torch.compile` needed) also crashes.** The bug is purely in `aten::conv2d`'s CPU path. Inductor is just one way to produce the triggering stride pattern:

```python
import torch

N, C, K, P = 2, 64, 125, 63
base = torch.randn(N, C, 1, 1)
x = torch.as_strided(base, (N, C, 1, 1), (C, 1, C, C))
assert torch.equal(x, base)

w = torch.randn(1, C, K, 1)
b = torch.randn(1)
torch.nn.functional.conv2d(x, w, b, stride=(2, 2), padding=(P, P))
```

**2. `gdb -batch -ex run -ex bt` resolves the fault to oneDNN's BRGEMM convolution kernel:**

```
dnnl::impl::cpu::x64::brgemm_convolution_fwd_t<...>::perform_outwork(...)
```

<details>
<summary>Full gdb backtrace</summary>

```
Thread 21 "python" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff50315b40 (LWP 153730)]
0x00007ffff0b8304d in dnnl::impl::cpu::x64::brgemm_convolution_fwd_t<(dnnl::impl::cpu::x64::cpu_isa_t)560>::perform_outwork(dnnl::impl::cpu::x64::brgemm_convolution_fwd_t<(dnnl::impl::cpu::x64::cpu_isa_t)560>::brgemm_thread_ctx_t const&, char*, char const*, int, int, bool, int, int, int, int, bool, bool, unsigned long, bool) const () from /home/harsh/projects/pytorch/.venv/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so
#0  0x00007ffff0b8304d in dnnl::impl::cpu::x64::brgemm_convolution_fwd_t<(dnnl::impl::cpu::x64::cpu_isa_t)560>::perform_outwork(...) const () from .../libtorch_cpu.so
#1  0x00007ffff0bdee1e in dnnl::impl::cpu::x64::brgemm_convolution_fwd_t<(dnnl::impl::cpu::x64::cpu_isa_t)560>::ker_base(...) const () from .../libtorch_cpu.so
#2  0x00007ffff0be00fb in dnnl::impl::cpu::x64::brgemm_convolution_fwd_t<(dnnl::impl::cpu::x64::cpu_isa_t)560>::execute(dnnl::impl::exec_ctx_t const&) const::{lambda(int, int)#1}::operator()(int, int) const () from .../libtorch_cpu.so
#3  0x00007ffff0b7c428 in dnnl::impl::parallel(int, std::function<void (int, int)> const&) [clone ._omp_fn.0] () from .../libtorch_cpu.so
#4  0x00007fffd91995d4 in ?? () from /usr/lib/libomp.so
#5  0x00007fffd91bc679 in __kmp_invoke_microtask () from /usr/lib/libomp.so
#6  0x00007fffd91034f4 in ?? () from /usr/lib/libomp.so
#7  0x00007fffd9107ea8 in ?? () from /usr/lib/libomp.so
#8  0x00007fffd91934a8 in ?? () from /usr/lib/libomp.so
#9  0x00007ffff7c97739 in ?? () from /usr/lib/libc.so.6
#10 0x00007ffff7d1bedc in ?? () from /usr/lib/libc.so.6
```

</details>

**3. Disabling mkldnn eliminates the crash entirely:**

```python
with torch.backends.mkldnn.flags(enabled=False):
    ...  # no crash
```

**4. Crash-triggering N differs from the original report** (N=2, 3, ... here vs N=3, 5, 6, 9, 10, 100 there), consistent with the allocation-layout-dependence already noted in the issue description but the underlying fault (same function, same conditions: non-contiguous strides on size-1 spatial dims + bias) is identical.

This looks like a bug in oneDNN's BRGEMM conv boundary-handling (`perform_outwork`) when given non-contiguous strides on size-1 spatial dims with bias. Likely worth cross-filing against oneDNN directly (uxlfoundation/oneDNN, formerly oneapi-src/oneDNN), since the fault is entirely inside `dnnl::` code linked into `libtorch_cpu.so`  though, I'll defer to the mkldnn integration owners here on whether it's a oneDNN-side bug or a PyTorch-side stride/memory-descriptor construction issue before filing upstream.

**Environment:**
- CPU: 13th Gen Intel(R) Core(TM) i5-13500HX
- torch built from source
- OS: Linux x86_64

**Comment by vpirogov** [comment]
Thanks for the report, @laolvfan. I can reproduce the issue directly in oneDNN and confirm it's reproducible on main branch. Created https://github.com/uxlfoundation/oneDNN/issues/5942 to track.
